Fixed cupy.mod for inf boundary condition

naxatra2 opened the pull request:
Fixes #8807

`cupy.mod` is just a variable alias for calling the `cupy.remainder` function. We would need to change the remainder function for the inclusion of inf.

asi1024 commented:
@naxatra2 Thank you for your contribution!
Currently, this pull request modifies the code for `cupy.array_api` module and does not fix the behavior of `cupy.remainder`. Could you kindly fix the `remainder` kernel definition in this location ([here](https://github.com/cupy/cupy/blob/0188dd8b16938fa835bcda797f70f9af2f8b4980/cupy/_core/_routines_math.pyx#L1075)) and add some tests in `tests/cupy_tests/math_tests/test_arithmetic.py`?

naxatra2 commented:
Sorry, I earlier got confused with the wrong code block.
But now I have done the changes that you suggested.
 

I updated that location and added some new test cases in `test_arithmetic.py`. 
I added thses tests: 
1. When the Divisor is zero (`x % 0`)  
2. When divisor is +ve or -ve inf (array case)  
3. When divisor is +ve or -ve inf (array case) 
4. When divisor ahs NaN values (`x % NaN`)  
5. when both are inf  (`inf % inf`).  

Also I explicitly suppressed the runtime warnings for undefined math results.

But one thing I noticed when I locally tested the file was that many other functions were raising errors even before I edited anything. 
<img width="958" alt="image" src="https://github.com/user-attachments/assets/e4d0602a-d266-49df-9a87-d3129677936c" />

asi1024 commented:
@naxatra2 
> But one thing I noticed when I locally tested the file was that many other functions were raising errors even before I edited anything.

I ran the test, but could not reproduce any error in my local environment. Could you show me the details of your environment obtained by `cupy.show_config()`?

naxatra2 commented:
>  Could you add tests for float16 and float32 by using @testing.for_float_dtypes() decorator?

I added the decorators for all the tests now.

naxatra2 commented:
> I ran the test, but could not reproduce any error in my local environment. Could you provide the details of your environment obtained by running 'cupy.show_config()'?

The commits of this PR are working fine. But I think there is some issue in my setup. I will try to reinstall my packages. 
```

OS                           : Windows-10-10.0.22631-SP0
Python Version               : 3.10.15
CuPy Version                 : 13.3.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.26.4
SciPy Version                : 1.14.1
Cython Build Version         : 0.29.36
Cython Runtime Version       : 3.0.11
CUDA Root                    : C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.6
nvcc PATH                    : C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.6\bin\nvcc.EXE
CUDA Build Version           : 12060
CUDA Driver Version          : 12060
CUDA Runtime Version         : 12060 (linked to CuPy) / 12060 (locally installed)
CUDA Extra Include Dirs      : []
cuBLAS Version               : (available)
cuFFT Version                : 11300
cuRAND Version               : 10307
cuSOLVER Version             : (11, 7, 1)
cuSPARSE Version             : (available)
NVRTC Version                : (12, 6)
Thrust Version               : 200600
CUB Build Version            : 200600
Jitify Build Version         : 1a0ca0e
cuDNN Build Version          : 8801
cuDNN Version                : 8907
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : NVIDIA GeForce GTX 1650 with Max-Q Design
Device 0 Compute Capability  : 75
Device 0 PCI Bus ID          : 0000:01:00.0
```

asi1024 commented:
@naxatra2 To run unit tests, you need to build the code from source in the main branch. Please refer to the following link for more details: https://docs.cupy.dev/en/stable/contribution.html#unit-testing

asi1024 commented on cupy/_core/_routines_math.pyx:
```diff
@@ -1071,15 +1071,21 @@ _floor_divide = create_ufunc(
-     ('dd->d', 'out0 = in0 - _floor_divide(in0, in1) * in1')),
-    'out0 = (in0 - _floor_divide(in0, in1) * in1) * (in1 != 0)',
+     ('ee->e',
+      'out0 = (in1 == INFINITY) ? in0 : '
```
The identifier "INFINITY" is undefined. Could you please fix the CUDA compilation error?

naxatra2 replied:
Can you please suggest some resources which might help me. I have tried doing it in various ways but still getting errors.

asi1024 replied:
Could you try `isinf`?
https://docs.nvidia.com/cuda/cuda-math-api/cuda_math_api/group__CUDA__MATH__SINGLE.html#_CPPv45isinff

naxatra2 commented:
> @naxatra2 To run unit tests, you need to build the code from source in the main branch. Please refer to the following link for more details: https://docs.cupy.dev/en/stable/contribution.html#unit-testing
Whenever I am trying to build it locally I am getting this error. 


Earlier too I was facing this same error, then I tried creating a new virtual environment and tried to install all the dependencies there, but still got the same error. I also cross verified the CUDA_PATH, it looks fine to me. I spent a lot of time trying to debug it but I can't understand where things are going wrong. Can you please help ?

![image](https://github.com/user-attachments/assets/d8c1dd56-5de4-4c34-b250-06348621880c)


---

Basically whenever I try to run :
`pip install -e .`

I get this error: 
```
    Environment Variables:
        CFLAGS          : (none)
        LDFLAGS         : (none)
        LIBRARY_PATH    : (none)
        CUDA_PATH       : C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.1
        NVCC            : (none)
        HIPCC           : (none)
        ROCM_HOME       : (none)
```

      Modules:
        cuda      : Yes (version 12010)
        cusolver  : Yes
        cudnn     : Yes (version 8907)
        nccl      : No
          -> Include files not found: ['nccl.h']
          -> Check your CFLAGS environment variable.
        nvtx      : Yes
        cutensor  : No
          -> Include files not found: ['cutensor.h']
          -> Check your CFLAGS environment variable.
        cub       : Yes (version 200001)
        jitify    : Yes (version 1a0ca0e)
        random    : Yes
        thrust    : Yes (version 200001)
        cusparselt: No
          -> Include files not found: ['cusparseLt.h']
          -> Check your CFLAGS environment variable.
        dlpack    : No
          -> Include files not found: ['cupy/_dlpack/dlpack.h']
          -> Check your CFLAGS environment variable.

      WARNING: Some modules could not be configured.
      CuPy will be installed without these modules.
      Please refer to the Installation Guide for details:
      https://docs.cupy.dev/en/stable/install.html

      ************************************************************

      Traceback (most recent call last):
        File "C:\Z\cupy-nax\mod-cupy\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
          main()
        File "C:\Z\cupy-nax\mod-cupy\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Z\cupy-nax\mod-cupy\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 181, in prepare_metadata_for_build_editable
          return hook(metadata_directory, config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\Nakshatra\AppData\Local\Temp\pip-build-env-81kl1l8v\overlay\Lib\site-packages\setuptools\build_meta.py", line 488, in prepare_metadata_for_build_editable
          return self.prepare_metadata_for_build_wheel(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\Nakshatra\AppData\Local\Temp\pip-build-env-81kl1l8v\overlay\Lib\site-packages\setuptools\build_meta.py", line 377, in prepare_metadata_for_build_wheel
          self.run_setup()
        File "C:\Users\Nakshatra\AppData\Local\Temp\pip-build-env-81kl1l8v\overlay\Lib\site-packages\setuptools\build_meta.py", line 522, in run_setup
          super().run_setup(setup_script=setup_script)
        File "C:\Users\Nakshatra\AppData\Local\Temp\pip-build-env-81kl1l8v\overlay\Lib\site-packages\setuptools\build_meta.py", line 320, in run_setup
          exec(code, locals())
        File "<string>", line 85, in <module>
        File "C:\Z\cupy-nax\install\cupy_builder\cupy_setup_build.py", line 532, in get_ext_modules
          extensions = make_extensions(ctx, compiler, use_cython)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Z\cupy-nax\install\cupy_builder\cupy_setup_build.py", line 377, in make_extensions
          raise Exception('Your CUDA environment is invalid. '
      Exception: Your CUDA environment is invalid. Please check above error log.
      [end of output]

```
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
(mod-cupy) PS C:\Z\cupy-nax> `
```

As the error suggested I tried reinstalling everything but its still not working

kmaehashi commented:
On Windows please follow the instructions here and do `git clone` again. Or simply use WSL for development.
https://docs.cupy.dev/en/latest/contribution.html#development-on-microsoft-windows

naxatra2 commented:
I am not able to do it. How can I define infinity ?? 
I tried doing this 
```
cdef extern from "<math.h>" :
    bint isinf(double x)
    bint isfinite(double x)
```
but it didn't work. also I tried to use infinity direcly from numpy but I am constantly getting errors. Can you please suggest from where can I learn it ?

kmaehashi commented:
We kindly request all contributors to test their pull requests on your local first. If you can't change the Windows setup, please use Linux for development instead.
